_adr/83/index.adoc

```diff
@@ -0,0 +1,132 @@
+. Make sure you have enough quota allocated. This requires:
+.. Having knowledge of the AMS endpoints and how to deal with them
+.. Having knowledge of the AMS terms used for designing the instance types (RHOSAKTrial for eval/developer and RHOSAK for standard)
+.. Enquire the quota assigned using the AMS endpoints, loop through the resources and check that enough quota is still available for product RHOSAK and resource RHOSAK
+.. If you have no quota assigned for STANDARD instances (RHOSAK), fallback to EVAL
+.. If you have quota for STANDARD (RHOSAK), but it is 0, fallback to EVAL (RHOSAKTrial)
+.. If you have quota for STANDARD (RHOSAK) and it is greater than 0 but all the granted quota is consumed, the creation should fail
+. Verify that at least one OCD Cluster has capacity for the instance you are going to create
+
+Another point of attention is that currently, any change in the flow requires an update to all the actors/consumers who are carrying out those checks.
+
+## Goals
+* API consumers can query KAS Fleet Manager Permission API to check if they can perform a certain action and receive a response with the outcome (allowed/denied) containing all the granted resources or, eventually, the missing requirements.
+
+## Non-goals
+
+## Current architecture
+* Kas Fleet Manager currently only allows the ‘try and see if it works’ approach: if what you are trying to do fails, Kas Fleet Manager gives you some details on went wrong, but you have no way of checking in advance if what you want to do is possible
+* API Consumers (like UI) are doing the required checkss into their codebase and every time a change occurs, both Kas Fleet Manager and UI have to be updated to reflect the change
+
+## Proposed architecture
+* We need to create a new API.
+The API will be authenticated and will allow the caller to check in advance whether he can perform a certain operation:
++
+```
+/can/{verb}/{resource}
+```
++
+Output of the API will be a JSON with the following structure:
++
+```
+{
+  allowed: boolean,
+  metadata: {
+    {key string}:{value}
+  }
+  reason: {
+    code: string,
+    description: string,
+    metadata: {             // additional details
+      {key string}: {value string}
+    }
+  }
+}
+```
++
+Some real world example for an invocation of the `/can/create/kafka` could be:
+
+.. Just one quota available: `standard.x2` won't be returned
++
+```
+{
+  allowed: true,
+  metadata: {
+    plans: ['standard.x1'],
+    quota: {
+      granted: 6,
+      used: 5
+    }
+  }
+}
+```
+
+.. More than one quota available:
++
+```
+{
+  allowed: true,
+  metadata: {
+    plans: ['standard.x1', 'standard.x2'],
+    quota: {
+      granted: 6,
+      used: 5
+    }
+  }
+}
+```
+
+.. Terms not accepted
++
+```
+{
+  allowed: false,
+  reason: {
+    code: "ERR-TERMS-01",
+    description: "Required terms have not been accepted",
+    metadata: {
+      acceptance_url: "https://url.to.accept.terms"
```

Comment:
Currently the only KFM error that comes to my mind as one needing to visit an URL is the `terms not accepted` error. That's why I put it into the `metadata` since I was seeing it as an exception

Reply:
Needing to buy more quote was the other example I was thinking of.

Reply:
Oh, I see now what you mean. 
Well, when the KFM invokes the terms API to know if the terms have been accepted or not, the API itself returns the url to be visited to fix the issue. The same does not happens for the AMS API (quota management).

I don't think KFM should store the URL, so I think it would be better to keep it into the `metadata` as a custom attribute.

Reply:
So the benefit of providing a URL is so that both the UI and CLI have this information, correct?

Regarding the use case @tombentley mentioned, we do have a Kafka overview page — https://console.redhat.com/application-services/streams/overview — that includes CTAs for getting or updating a subscription. Eventually, pricing information will live on redhat.com, so the design will change when that happens. Maybe we would have a single CTA on the Kafka overview page related to getting or updating a subscription that points to the page on redhat.com 🤷‍♀️ I like the idea of providing this URL, as Tom suggests, but I don't know exactly how these pages and contents will evolve.